### PR TITLE
Support for running NTO e2e tests.

### PR DESCRIPTION
**Pull request by jmencak**
This is the first attempt at leaveraging this repo to run NTO e2e tests.

*Examples:*
Run NTO's e2e tests automatically based on `clusterversion`.
```
run nto e2e
```
Run NTO's e2e 4.7 tests.
```
run nto e2e origin/release-4.7
```

Run NTO's e2e tests for a specific NTO commit.
```
run nto e2e 14bdf4d4
```

This is the first step.  Ideas how to improve or approach this in a different way more than welcome.  Thanks!

### CODE REVIEW

**Comment by kpouget** [comment]
Hello @jmencak, thanks for this first PR on this repo ;-)

the way we usually organize the code/entry path is:
1. `ci_entrypoint_nto.sh` calls a `toolbox` script, eg `toolbox/nto/test-e2e.sh`
2. `toolbox/nto/test-e2e.sh` prepare the test parameters (here `commit=origin/release-$OCP_VERSION`) and calls ansible
3. ansible performs the test steps (here, `git clone` + `make test`)

I would suggest that you follow these 3 levels, or at least 1 and 2, for consistency of the repo organization.

I think you (we) would benefit from following 3 as well, to have the logs properly organized.
By calling an ansible playbook, you'll get a dedicated directory for the test artifacts + a parsable log file of the playbook tasks:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-psap-ci-artifacts-release-4.7-gpu-operator-e2e-master/1401675477935984640/artifacts/gpu-operator-e2e-master/nightly/artifacts/002__entitlement__deploy/

it's hard to say if it's really necessary for running 2 tasks, what do you think?
and we can already make it grow 50% bigger by adding a task storing the current git commit, to know immediately what was tested:
```
git show --quiet > {{ artifact_extra_logs_dir }}/git_commit
```

you can also call `toolbox/cluster/capture_environment.sh` to store the OCP version + `ci-artifacts` commit, again to know what was tested

**Comment by jmencak** [comment]
Thank you for the valuable input, Kevin!  Hopefully adjusted closer to what you expect.  Those three "levels" seem overkill at first, but makes sense on second look.  If I missed any of your comments, happy to adjust.

File: build/root/usr/local/bin/ci_entrypoint_nto.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        exit 0
+        ;;
+    -*)
+        echo "FATAL: Unknown option: ${action}"
+        exit 1
+        ;;
+    *)
+        echo "FATAL: Unknown action \"${action}\""
+        exit 1
+        ;;
```
Comment (kpouget) [inline build/root/usr/local/bin/ci_entrypoint_nto.sh:34]: do you know if there is a difference between these two cases?
`*)` is the default, but `-*)` I'm not sure

Comment (jmencak) [reply build/root/usr/local/bin/ci_entrypoint_nto.sh:34]: Not sure I understand the question.  I added the `\"${action}\"` to assist better debugging.  "FATAL: Nothing to do ..." doesn't help much.

`-*)` will catch the actions starting with `-` (i.e. options),
`*)` will catch the rest (empty string, ...).

and at least you'll have a better idea what happened.

Comment (kpouget) [reply build/root/usr/local/bin/ci_entrypoint_nto.sh:34]: oh, I though `-*)` was a special bash syntax :# :# :#

now I understand better, I thought my `nothing to do` would only be executed when no parameter was passed, that's why I didn't print the argument value in the message...

```
    -*)
            echo "FATAL: Unknown option: ${target}"     
            exit 1
```
this block is not necessary then, we never expect "options" in the entrypoints, at least in `run`

File: build/root/usr/local/bin/run [hunk 2]
```diff
@@ -121,7 +125,7 @@ case ${target} in
         exit 1
         ;;
     *)
-	echo "FATAL: Nothing to do ..."
+	echo "FATAL: Unknown target \"${target}\""
```
[the remark this answers is not in the document]
Comment (kpouget) [reply build/root/usr/local/bin/run:128]: copy and paste of the message in the other comment, as it more belongs here:

```
    -*)
            echo "FATAL: Unknown option: ${target}"     
            exit 1
```
this block is not necessary then, we never expect "options" in the entrypoints

could you also realign the `echo`, it seems it's 1 level too low

File: build/root/usr/local/bin/ci_entrypoint_nto.sh (hunk 1 above)
Comment (jmencak) [reply build/root/usr/local/bin/ci_entrypoint_nto.sh:34]: OK, will remove the option's block then. :)

**Comment by jmencak** [comment]
> Also, quick question, why did you decide to use `ci-artifacts` instead of directly using NTO repo as the base of the testing?
> when I prepared the Prow [PR](https://github.com/openshift/release/pull/17048) it made sense to me not to use `ci-artifacts`, as you have only one toolbox command here. The main point of the repo is to orchestrate multiple "complex" steps required for the testing, while with NTO, you have only one "simple" (= `make test`) step
> 
> anyway, when we'll progress with the `ci-dashboard`, you'll have additional benefits for free ;-)

That's the plan. Some of the reasons I wanted to use the `ci-artifacts`:
- I wanted to keep the `openshift/release` for NTO as simple as possible given the issues we had with the first version of https://github.com/openshift/release/pull/17020
- I wanted to test NTO on nightly/ART builds.  I do not believe this can be easily done with openshift/release as is at the moment as there are still pending some ART Jira tickets for that
- You've made a lot of progress on the `ci-artifacts/ci-dashboard` repos; the dashboards look cool and I'd like to leaverage that for NTO too in the future, as you say nearly "for free".

**Comment by kpouget** [comment]
ack,

> I wanted to keep the openshift/release for NTO as simple as possible given the issues we had with the first version of openshift/release#17020

indeed, makes sense

> I wanted to test NTO on nightly/ART builds. I do not believe this can be easily done with openshift/release as is at the moment as there are still pending some ART Jira tickets for that

FYI, for what I see in the current code of the PR, using openshift/release would test the exact same thing
`ci-artifacts` is just a man in the middle, but Prow 1/ clones a git repo 2/ triggers a command side, that's what we have here

> You've made a lot of progress on the ci-artifacts/ci-dashboard repos; the dashboards look cool and I'd like to leaverage that for NTO too in the future, as you say nearly "for free".

:-)
FYI again, the current dashboard works with any Prow test, I only parse the file-artifacts generated by Prow
but in the future, I would like to take into account a) the toolbox steps (`000_`, `001_`) to show which step step failed,
b) the `_ansible_logs.json`, first to show "X test succeeded, Y test skipped, Z test failed", and maybe also render the logs task by task

**Comment by jmencak** [comment]
This should be hopefully much closer to the final PR, Kevin, right? :)

File: roles/nto_run_e2e_test/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+- name: Git checkout the NTO repository with e2e tests
+  ansible.builtin.git:
+    repo: "{{ nto_git_repo }}"
+    dest: "{{ nto_tmp_dir }}"
+    version: "{{ nto_git_ref }}"
+    force: true
+
+- name: Store the current git commit
+  shell:
```
Comment (kpouget) [inline roles/nto_run_e2e_test/tasks/main.yml:18]: we need to turn this error into a warning

Comment (jmencak) [reply roles/nto_run_e2e_test/tasks/main.yml:18]: How?

Comment (jmencak) [reply roles/nto_run_e2e_test/tasks/main.yml:18]: And actually, do we even need `git show -s ` ?

Comment (kpouget) [reply roles/nto_run_e2e_test/tasks/main.yml:18]: you can mute it with this task tag (I just discovered it):
```
+  tags: [skip_ansible_lint]
```

> And actually, do we even need git show -s ?

yes, to keep a parsable trace of what is being tested
it's already part of the logs of the task above, but with a dedicated artifact file, it's simpler to confirm what state of the repo is being tested.
eg, the test failed, but I'm sure I fixed it today --> look at the commit, find out if your patch was actually taken into account in the nightly testing

**Comment by jmencak** [comment]
`OutOfMemoryKilled: The build pod was killed due to an out of memory condition.`
/test sro-images

**Comment by kpouget** [comment]
@jmencak , I tested locally, good news is that the test ran, bad news is that the test failed!
```
• Failure [305.058 seconds]
[basic][rollback] Node Tuning Operator settings rollback
/tmp/nto/test/e2e/basic/rollback.go:18
  Tuned settings rollback
  /tmp/nto/test/e2e/basic/rollback.go:26
    net.ipv4.tcp_tw_reuse set [It]
    /tmp/nto/test/e2e/basic/rollback.go:40

    Unexpected error:
        <*errors.errorString | 0xc0006d21e0>: {
            s: "sysctl net.ipv4.tcp_tw_reuse=2 in Pod tuned-vx7bx, expected 1: command [sysctl -n net.ipv4.tcp_tw_reuse] outputs 2 (leading/trailing whitespace trimmed) in Pod tuned-vx7bx, expected 1: <nil>",
        }
        sysctl net.ipv4.tcp_tw_reuse=2 in Pod tuned-vx7bx, expected 1: command [sysctl -n net.ipv4.tcp_tw_reuse] outputs 2 (leading/trailing whitespace trimmed) in Pod tuned-vx7bx, expected 1: <nil>
    occurred

    /tmp/nto/test/e2e/basic/rollback.go:71
------------------------------
SSSSSSSS

Summarizing 1 Failure:

[Fail] [basic][rollback] Node Tuning Operator settings rollback Tuned settings rollback [It] net.ipv4.tcp_tw_reuse set
/tmp/nto/test/e2e/basic/rollback.go:71

Ran 1 of 9 Specs in 305.058 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 8 Skipped
--- FAIL: TestNodeTuningOperator (305.06s)
FAIL
FAIL	github.com/openshift/cluster-node-tuning-operator/test/e2e/basic	305.073s
FAIL
make: *** [Makefile:91: test-e2e] Error 1
```
apart from that, looks good !

**Comment by jmencak** [comment]
> @jmencak , I tested locally, good news is that the test ran, bad news is that the test failed!
> 
> ```
> • Failure [305.058 seconds]
> [basic][rollback] Node Tuning Operator settings rollback
> /tmp/nto/test/e2e/basic/rollback.go:18
>   Tuned settings rollback
>   /tmp/nto/test/e2e/basic/rollback.go:26
>     net.ipv4.tcp_tw_reuse set [It]
>     /tmp/nto/test/e2e/basic/rollback.go:40
> 
>     Unexpected error:
>         <*errors.errorString | 0xc0006d21e0>: {
>             s: "sysctl net.ipv4.tcp_tw_reuse=2 in Pod tuned-vx7bx, expected 1: command [sysctl -n net.ipv4.tcp_tw_reuse] outputs 2 (leading/trailing whitespace trimmed) in Pod tuned-vx7bx, expected 1: <nil>",
>         }
>         sysctl net.ipv4.tcp_tw_reuse=2 in Pod tuned-vx7bx, expected 1: command [sysctl -n net.ipv4.tcp_tw_reuse] outputs 2 (leading/trailing whitespace trimmed) in Pod tuned-vx7bx, expected 1: <nil>
>     occurred
> 
>     /tmp/nto/test/e2e/basic/rollback.go:71
> ------------------------------
> SSSSSSSS
> 
> Summarizing 1 Failure:
> 
> [Fail] [basic][rollback] Node Tuning Operator settings rollback Tuned settings rollback [It] net.ipv4.tcp_tw_reuse set
> /tmp/nto/test/e2e/basic/rollback.go:71
> 
> Ran 1 of 9 Specs in 305.058 seconds
> FAIL! -- 0 Passed | 1 Failed | 0 Pending | 8 Skipped
> --- FAIL: TestNodeTuningOperator (305.06s)
> FAIL
> FAIL	github.com/openshift/cluster-node-tuning-operator/test/e2e/basic	305.073s
> FAIL
> make: *** [Makefile:91: test-e2e] Error 1
> ```
> 
> apart from that, looks good !

Interesting, tested locally 3 times, no failures. :)

**Comment by kpouget** [comment]
thanks @jmencak looks good :D 

\lgtm

**Comment by kpouget** [comment]
\approve

**Comment by kpouget** [comment]
thanks @jmencak looks good :D

/lgtm
/approve

**Pull request merged**